### ISSUE TITLE
Disallow operations on scalar tensors that are no-ops

### ISSUE THREAD

**Issue by reillyeon** [webmachinelearning/webnn#794]
There are a number of operators, such as `transpose()` and `reshape()` which are no-ops when applied to a scalar tensor. In addition there are operators like `gather()` which expect indices but a scalar doesn't have indices.

While implementations can work around this (typically by treating a scalar as a single-element 1D tensor) it seems better to simply disallow these cases which add complexity without adding an expressivity.

**Comment by reillyeon** [comment]
As an additional data point, both LiteRT and Core ML do not directly support applying their transpose operators to a scalar and implementations based on these frameworks will require workarounds. DirectML only supports this operation because transposition is implemented by adjusting the dimensions and strides of a tensor as it is passed between other nodes, which is a no-op for a scalar tensor.

**Comment by reillyeon** [comment]
We recall that @fdwr had a reason why supporting this was useful for developers, but if so then we should include a more complete discussion of tensor operations on scalars in the specification. Right now there's not much mention that a tensor can have a rank of 0.

**Comment by fdwr** [comment]
I have no qualms about disallowing this for `gather`, and I can contently let go of `transpose` (because one must balance between *"what logically should happen to follow consistent semantics and is mathematically sensible"* vs *"what is practical to implement in the real world"*), but `reshape` as a more fundamental operation I really feel should follow the axiom that you can always reshape a tensor to its own shape. So if you have an input of `sizes = [4,3]`, then it should be legal to reshape it to `sizes = [4,3]`; and if you have `sizes = []`, then you should be able to reshape it to `sizes = []` - a consistent rule, orthogonal to scalarity or not. Granted, it *seems* useless at first, but it does happen in higher levels with generic routines that you encounter these kinds of cases.

I don't have easy access to CoreML or TFLiteRT to try, but I'd be surprised if `reshape` of a scalar didn't work as expected for them, as TF and PT and NumPy all agree (code at bottom)?

> Right now there's not much mention that a tensor can have a rank of 0.

Yeah, we could add more mention. [`reshape`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-reshape-method) is documented to support rank `N` (thanks to @inexorabletash's https://github.com/webmachinelearning/webnn/pull/657), which means 0 through the maximum backend limit  ...

operand	| allowed data types	|allowed ranks
-----------|---------------------------|--------------
input	|any	| <u><b>N</b></u>
output	|same as input	|newShape's size

...but we could clarify the existing wording "allowed ranks ... are given as an explicit rank (e.g. 1), or N to allow any dimensionality" to explicitly include `0` when defining `N`. Note the general expectation is that *most* operators support scalars by default, except for those operators that don't (rather than treating scalars as a special case that only apply to a subset of operators).

## Reshape reference code

```python
# pip freeze torch==1.11.0+cpu
import tensorflow as tf

x = tf.constant(42, dtype=tf.float32)
y = tf.reshape(x, [])

print("value:", y)
print("shape:", y.shape)

# value: 42
# shape: []
```

```python
# pip freeze torch==1.11.0+cpu
import torch

x = torch.tensor(42, dtype=torch.float32)
print("x value:", x)
print("x shape:", x.shape)
print("x dtype:", x.dtype)

y = x.reshape(())
print("y value:", y)
print("y shape:", y.shape)
print("y dtype:", y.dtype)

# x value: tensor(42.)
# x shape: torch.Size([])
# x dtype: torch.float32
# y value: tensor(42.)
# y shape: torch.Size([])
# y dtype: torch.float32
```

```python
# pip freeze numpy==1.21.6
import numpy
x = numpy.array(42, dtype=numpy.float32)
y = x.reshape([])

print("value:", y)
print("shape:", y.shape)
print("dtype:", y.dtype)

# value: 42.0
# shape: ()
# dtype: float32
```

**Comment by reillyeon** [comment]
I agree with the argument in favor of `reshape()`. It was incorrect to refer to it as a no-op since its purpose is to change the shape, rather than the value, of the tensor and changing the shape from a scalar to a single-element tensor is valid.

I'm less sure about other operators. For example `gather()` probably makes sense if you require that `indices` is not a scalar as otherwise the output rank calculation would return -1. `pad()` only makes sense if you implicitly reshape the input to a single-element tensor first.

What I'd like to see are examples of where silently tolerating scalars is useful for developers. This might end up in a similar place as #391.

To start we should go through each case where we allow a rank of N and clarify whether 0 is a valid value of N.

**Comment by inexorabletash** [comment]
> we could clarify the existing wording "allowed ranks ... are given as an explicit rank (e.g. 1), or N to allow any dimensionality" to explicitly include `0` when defining `N`

+1 - if this issue turns into a PR we should include that.

**Comment by fdwr** [comment]
Next steps:
- [x] Enumerate all operators, and determine which:
  - ✅ support scalars
  - ❌ definitely do not (any requiring a rank greater than 0 or an `axis` parameter are not applicable anyway)
  - ❔ are in question here
- [ ] For ❔,  determine which backends support scalars already, which do not, and whether it's worth extra code to wrap that operator in a temporary reshape to 1D (this already happens in the DML backend because the tensor description is massaged to 1D before calling DML, but it may require actual explicit `reshape` calls in other backends).
  - Note that frameworks might generate some of these constructs as a by-product, even if a user directly calling the API would probably avoid doing so, and so that should be considered too.
  - Shape manipulation operators (like `reshape` and `expand`) have higher priority to me to support scalars, since those are more likely to occur as generation by products.

| Operator                           | Verdict |
|------------------------------------|---------|
| argMin/argMax operations           | ❌ requires axis |
| batchNormalization                 | ✅ input.rank == 0 to N |
| cast                               | ✅ |
| clamp                              | ✅ |
| concat                             | ❌ requires axis |
| conv2d                             | ❌ input.rank == 4 |
| convTranspose2d                    | ❌ input.rank == 4 |
| Element-wise binary operations     | ✅ |
| Element-wise logical operations    | ✅ |
| Element-wise unary operations      | ✅ |
| elu                                | ✅ |
| expand                             | ❔ consistent with reshape |
| gather                             | ❔ |
| gelu                               | ✅ |
| gemm                               | ❌ input.rank == 2 |
| gru                                | ❌ input.rank == 3 |
| gruCell                            | ❌ input.rank == 2 |
| hardSigmoid                        | ✅ |
| hardSwish                          | ✅ |
| instanceNormalization              | ❌ input.rank == 4 |
| layerNormalization                 | ✅ input.rank == 0 to N |
| leakyRelu                          | ✅ |
| linear                             | ✅ |
| lstm                               | ❌ input.rank == 3 |
| lstmCell                           | ❌ input.rank == 2 |
| matmul                             | ❌ input.rank >= 2 |
| pad                                | ❔ |
| Pooling operations | ❌ input.rank == 4 |
| prelu                              | ✅ |
| Reduction operations               | ✅ see [`noop_with_empty_axes`](https://onnx.ai/onnx/operators/onnx__ReduceMean.html) for nop cases |
| relu                               | ✅ |
| resample2d                         | ❌ input.rank == 4 |
| reshape                            | ❔ |
| sigmoid                            | ✅ |
| slice                              | ❔ |
| softmax                            | ❌ requires axis |
| softplus                           | ✅ |
| softsign                           | ✅ |
| split                              | ❌ requires axis |
| tanh                               | ✅ |
| transpose                          | ❔ |
| triangular                         | ❌ input.rank >= 2 |
| where                              | ✅ |

That leaves:

| Operator                           | Verdict |
|------------------------------------|---------|
| expand                             | ❔ dwayner - should be consistent with reshape |
| gather                             | ❔ |
| pad                                | ❔ |
| reshape                            | ❔ dwayner - my POV is yes ✅ |
| slice                              | ❔ |
| transpose                          | ❔dwayner - logically the backends should just treat these as no-axis nops, but it may not be worth the added code to enable that in the WebNN calling layer |

**Comment by huningxin** [comment]
Re Chromium prototype:

- [expand](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.cc;l=2116?q=ml_graph_builder.cc&ss=chromium%2Fchromium%2Fsrc) supports scalar.
- [gather](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=1329) doesn't support scalar. It has `MLGatherOptions.axis`. 
- [pad](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=931) doesn't support scalar. It may be able to support scalar if empty beginningPadding and endingPadding mean nop.
- [reshape](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.cc;l=2793?q=ml_graph_builder.cc&ss=chromium%2Fchromium%2Fsrc) supports scalar.
- [slice](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=2405) doesn't support scalar. It may be able to support scalar if empty starts, ends and strides mean nop.
- [transpose](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;drc=3b5e9130048e66ce39338d34f99751d0a2346e9c;l=2343) allows scalar, empty permutation is a nop.

**Comment by fdwr** [comment]
> gather doesn't support scalar. It has MLGatherOptions.axis.

@huningxin : TY, missed that when filling in table. Updated.

**Comment by huningxin** [comment]
For other wave3 gather/scatter ops, according to Chromium prototype:
* [gatherElements](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=1386;bpv=1;bpt=0) doesn't support scalar `input` and `indices`. It requires `axis` and `indices.rank` == `input.rank`.
* [gatherND](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=1446;bpv=1;bpt=0) doesn't support scalar `input` and `indices`. According to its definition, the last dimension of `indices` contains indices of elements or slices in `input` tensor. Both should be at least 1D.
* [scatterElements](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=2574;bpv=1;bpt=0) doesn't support scalar `input` and `indices`. Same as `gatherElements`.
* [scatterND](https://source.chromium.org/chromium/chromium/src/+/main:services/webnn/public/cpp/graph_validation_utils.cc;l=2643;bpv=1;bpt=0) doesn't support scalar `input` and `indices`. Same as `gatherND`.
